Add dynamically trained MADWF blocking to MSPCG

hummingtree opened the pull request:
This PR adds an exploratory feature that enables using a Mobius DWF operator with reduced `Ls` for MSPCG.

The idea aligns with the M\"obius Accelerated Domain Wall Fermion (MADWF, [arXiv:1111.5059](https://arxiv.org/abs/1111.5059)): having a M\"obius operator with reduced `Ls` that is *close* to the original operator to speed up inversions. The original MADWF algorithm is based on the domain wall - overlap transformation to transform between different `Ls`. In principle for MSPCG we can use the same trick for the preconditioner, but the application of MADWF will not be fast enough due to the multiple inversions involved.

Instead we apply a simple transfer matrix to transform between different `Ls`. For the preconditioning inversion, the rhs is first transformed to reduced `Ls`, after which a inversion is performed. The solution, which lives in the reduce `Ls` space, is transformed back into the original `Ls` space with the very same transfer matrix.

And the transfer matrix is *trained* with a straight forward loss function.

See section 3.3 of [arXiv:2104.05615](https://arxiv.org/abs/2104.05615).

See the [wiki page](https://github.com/lattice/quda/wiki/Preconditioning) for more info.

A few remaining TODOs:
- [x] Remove the duplicate (alternative) reliable update routines in both PCG and CG: create a reliable update `struct` to handle this.
- [x] Add issue for creating preconditioning solvers.
- [x] Update corresponding sections of wiki.
- [x] Clang-format.

### `16` nodes, grid layout `2x2x4x6`
***
- CG, double-half
```
6 solves, mean iteration count 11383.4 (stddev = 31.1095), with mean solve time 217.133 (stddev = 0.908744), mean GFLOPS 106630 (stddev = 170.277) [excluding first solve]
```

- MSPCG, suppressor = `0.848`, `double-half-half`, `6` inner iteration, `Ls`: `12->4`
```
6 solves, mean iteration count 4778 (stddev = 23.5266), with mean solve time 172.357 (stddev = 0.927253), mean GFLOPS 169292 (stddev = 180.042) [excluding first solve]
```
`speed up = 217.133/172.357 = 1.26`

### `64` nodes, grid layout `4x4x4x6`
***
- CG, double-half
```
6 solves, mean iteration count 11384.2 (stddev = 30.9629), with mean solve time 92.531 (stddev = 0.766309), mean GFLOPS 250247 (stddev = 1464.39) [excluding first solve]
```
- MSPCG, suppressor = `0.850`, `double-half-half`, `6` inner iteration, `Ls`: `12->4`
```
6 solves, mean iteration count 5312.4 (stddev = 39.9287), with mean solve time 70.6399 (stddev = 0.432256), mean GFLOPS 491310 (stddev = 736.662) [excluding first solve]
```
`speed up = 92.531/70.6399 = 1.31`

An example `invert_test` cmd options: (note the `additive-madwf` option for `schwarz-type` and all the options starting with `madwf-`)
```
 --dslash-type mobius --dim ${LOCALSIZE} --gridsize ${GRIDSIZE} \
 --load-gauge ${GAUGE} \
 --Lsdim 12 --m5 -1.1 --b5 1.5 --c5 0.5 --mass 0.00152 \
 --verbosity verbose --nsrc 6 \
 --solution-type mat --matpc even-even --solve-type normop-pc \
 --prec double --prec-sloppy half --reliable-delta 0.1 --prec-precondition half \
 --recon 12 \
 --alternative-reliable true \
 --inv-type pcg --precon-type cg --precon-schwarz-type additive --precon-accelerator-type madwf \
 --tol 1e-10 --tol-precondition 1e-6 --niter 14000 \
 --maxiter-precondition 6 \
 --madwf-diagonal-suppressor 0.850 \
 --madwf-ls 4 \
 --madwf-null-miniter 100 \
 --madwf-null-tol 1e-6 \
 --madwf-train-maxiter 600 \
 --madwf-param-load true \
 --madwf-param-save false \
 --madwf-param-infile ${QUDA_RESOURCE_PATH} \
 --madwf-param-outfile ${QUDA_RESOURCE_PATH}
```

The formal added paramerter to `inv_param`
```
    /**
     * The following parameters are the ones used to perform the adaptive MADWF in MSPCG
     * See section 3.3 of [arXiv:2104.05615]
     */

    /** The diagonal constant to suppress the low modes when performing 5D transfer */
    double madwf_diagonal_suppressor;

    /** The target MADWF Ls to be used in the accelerator */
    int madwf_ls;

    /** The minimum number of iterations after which to generate the null vectors for MADWF */
    int madwf_null_miniter;

    /** The maximum tolerance after which to generate the null vectors for MADWF */
    double madwf_null_tol;

    /** The maximum number of iterations for the training iterations */
    int madwf_train_maxiter;

    /** Whether to load the MADWF parameters from the file system */
    QudaBoolean madwf_param_load;

    /** Whether to save the MADWF parameters to the file system */
    QudaBoolean madwf_param_save;

    /** Path to load from the file system */
    char madwf_param_infile[256];

    /** Path to save to the file system */
    char madwf_param_outfile[256];

```

This PR also fixes a few issues and adds a few features: 
- `invert_test` now correctly initialize random rhs with `ndim = 5`. (Previously it assumes `ndim = 4`.)
- Abstract the reliable update (whether alternative or not) logics into a separate class.
- Use the class to add alternative reliable update to preconditioned CG.
- Use Polak-Ribière version of the beta update in preconditioned CG.

hummingtree committed:
Half way through adding MSPCG-ML: NOT YET WORKING!

hummingtree committed:
Add MADWF-ML.

hummingtree committed:
Clang.

hummingtree committed:
Changes to the interface:
- Change file names.
- Add file system load and save routines.
- Fix thrust include problem.

hummingtree committed:
Add todo's to madwf_ml.h. Remove old interfaces to Solver.

hummingtree committed:
Add null vector options.

hummingtree committed:
Add alternative reliable update to PCG.

hummingtree committed:
Add PR beta calculation even if a reliable update is performed.

hummingtree committed:
Clang.

hummingtree committed:
Save tunecache after generating the null vectors.

hummingtree committed:
Some desparate changes for speed up:
- Turn off unnecessary advanced features of CG if we run MSPCG
- Turn on quarter precision for the accerlated preconditioner.

hummingtree committed:
Fix the cache for trained parameters. Time the loading time.

hummingtree committed:
Not profile when training to avoid timer crashes.

hummingtree committed:
Use gaussian noise vector for the training - suspecting that's the reason why training fails.

hummingtree committed:
Fix how invert_test generate random rhs: previously it simply ignores the fifth dimension all together.

hummingtree committed:
Use customized cubic solver. Recover the code that enables tuning mu.

hummingtree committed:
Change my affiliation.

hummingtree committed:
Add the option to set max_res_increase(_total).

hummingtree committed:
extern variables should not be initialized.

hummingtree committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/mspcg-ml

hummingtree committed:
Cleanning up the code. Move transfer and tensor to different files.

hummingtree committed:
More cleaning up: fix the terminologies so they make sense. Make `solve_and_collect` a method for PCG.

hummingtree committed:
Add comments to quda.h

hummingtree committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/mspcg-ml

hummingtree committed:
Add madwf_tensor.cu

hummingtree committed:
Fix a merge bug so hq_res_stall_check is used.

hummingtree committed:
Use cub for the block reduce and warp reduce.

hummingtree committed:
Put training inside the init profile timer.

hummingtree committed:
Address clang compiler warnings.

hummingtree committed:
Remove redundant code. Move init profile after the training.

hummingtree committed:
Backup buffer data to avoid pollution resulting from tuning.

hummingtree committed:
Remove unused variable.

hummingtree committed:
Only add __CUDA_ARCH__ to only necessary lines.

hummingtree committed:
Move the reliable update code in CG and PCG to a common struct called reliable_updates.

hummingtree committed:
Add the missing include reliable_updates.h

hummingtree committed:
Use QUDA's cub wrapper for block reduce and warp reduce, as well as static shared memory wrapper.

hummingtree committed:
Apply clang-format to the new files.

hummingtree committed:
Fix a long time bug where the block reduce is not performed correctly. This bug existed from the beginning but was not exposed due to block size = warp size.

hummingtree committed:
Apply clang-format to devcie_vector.cu

hummingtree committed:
Fix check_param.cpp. Add doc in shared_memory_cache_helper.cuh.

hummingtree committed:
Add comments to the madwf related params in invert_quda.h. Fix check_param.h.

hummingtree committed:
Merge remote-tracking branch 'origin/feature/cub_abstraction' into feature/mspcg-ml

hummingtree committed:
Various misc fixes:
- Remove the FIXME and address the issue by always checking if only single parity
- Move the DWF macro to outside the function to suppress compiler warnings
- Add flops count for the tensor and transfer

hummingtree committed:
Merge remote-tracking branch 'origin/develop' into feature/mspcg-ml

hummingtree committed:
Merge remote-tracking branch 'origin/develop' into feature/mspcg-ml

maddyscientist commented on include/invert_quda.h:
```diff
@@ -245,6 +245,37 @@ namespace quda {
         interface)*/
     bool mg_instance;
 
+    /** The diagonal constant to suppress the low modes when performing 5D transfer */
```
These parameters also need to be mirrored to quda_fortran.F90

maddyscientist commented on include/quda.h:
```diff
@@ -334,6 +334,38 @@ extern "C" {
     /** Whether to use additive or multiplicative Schwarz preconditioning */
     QudaSchwarzType schwarz_type;
 
+    /**
```
These parameters need to be mirrored to quda_fortran.F90

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/2769f6268cd7367667fc6c2b38cde0a44442f28c. Can you take a look and make sure it works with fortran?

maddyscientist replied:
I've verified that the Fortran interface is working fine with this change.  Thanks.

maddyscientist commented on include/polynomial.h:
```diff
@@ -0,0 +1,114 @@
+#include <array>
+#include <vector>
+
+// Implimented according to https://en.wikipedia.org/wiki/Cubic_equation
```
typo

maddyscientist commented on include/polynomial.h:
```diff
@@ -0,0 +1,114 @@
+namespace quda
+{
+
+  constexpr double PI = 3.141592653589793;
```
Can just use `M_PI` (provided by the math library)

hummingtree replied:
`M_PI` does seem to be present in all compilers? I did a search and I do not find it in the C++ standard.

maddyscientist replied:
I learned something new here, that `M_PI` is not part of math.h / cmath, and is compiler dependent.  It's definitely supported in all compilers that QUDA is used by, and we do utilize `M_PI` throughout QUDA.

If one is concerned with using `M_PI`, we could add the following to `quda_constant.h`
```c++
#ifndef M_PI
#define M_PI (3.14159265358979323846264338327950288)
#endif /* M_PI */
```

hummingtree replied:
Personally I am fine with using `M_PI` directly but I think the above solution is better. :)

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/d776b593adac861010d314b14523ba261ada3c14

maddyscientist commented on include/device_vector.h:
```diff
@@ -0,0 +1,90 @@
+#include <malloc_quda.h>
+#include <util_quda.h>
+#include <complex_quda.h>
+
```
This file should be properly documented with doxygen.  This container looks like it could be useful in general.

maddyscientist commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,205 @@
+#include <block_reduce_helper.h>
+
+namespace quda
+{
```
Can you add some documentation to this file?  A general description of what the kernel does here would be welcome.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/cefb2a527fa7cb1303a9e7487b9fea4f39d1bf8d

maddyscientist commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,205 @@
+
+  namespace madwf_ml
+  {
+    template <class T> __device__ __host__ inline void block_reduce_x(T &f)
```
What does `block_reduce_x` do that makes it distinct from `BlockReduce`?  If you are only wanting to do a reduction in the x thread dimension, we could likely modify the "batched" `BlockReduce` to meet the needs of this.  It is designed to do distinct reductions across the block z dimension and not the y dimension like is used here, but we could modify it accordingly easily enough.  The only place where the batched reductions are done is in the restrictor.

hummingtree replied:
Indeed we are just performing `BlockReduce` in the x direction.

maddyscientist commented on include/invert_quda.h:
```diff
@@ -245,6 +245,37 @@ namespace quda {
         interface)*/
     bool mg_instance;
 
+    /** The diagonal constant to suppress the low modes when performing 5D transfer */
```
There's a lot of parameters here specifically for MADWF.  It would make it cleaner if we have a container that packs all these in.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/e5cf6f4989bb7a3c71a0d6b1d11b9348b8afdf38

maddyscientist commented on lib/madwf_transfer.cu:
```diff
@@ -0,0 +1,97 @@
+      if (tp.size() * sizeof(float) != m_size) {
+        errorQuda("Training Parameter size mismatch %lu neq %lu.\n", tp.size() * sizeof(float), m_size);
+      }
+
```
Can this be written using `instantiatePreconditioner` or similar to allow us to only compile which precisions are enabled in CMake?

hummingtree replied:
~~Only `float` is instantiated here, and we only need `float`, so I guess there is no need to use something like `instantiatePreconditioner`?~~

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/ff335265b55f103ee3927abdf7ef5671034dade2

maddyscientist commented on lib/madwf_transfer.cu:
```diff
@@ -0,0 +1,97 @@
+      {
+        TunableKernel2D_base<false>::resizeStep(out.X(4)); // Ls must be contained in the block
+
+        strcpy(aux, out.AuxString());
```
This line should be deleted, since the parent `TunableKernel` class will already initialize the `aux` string with `out.AuxString()`

maddyscientist commented on lib/madwf_tensor.cu:
```diff
@@ -0,0 +1,123 @@
+      {
+        TunableKernel2D_base<false>::resizeStep(out.X(4)); // Ls must be contained in the block
+
+        strcpy(aux, out.AuxString());
```
Can delete this line, since `aux` will be set accordingly in the the `TunableKernel` parent

maddyscientist commented on lib/madwf_tensor.cu:
```diff
@@ -0,0 +1,123 @@
+        errorQuda("Training Parameter size mismatch %lu neq %lu.\n", transfer_parameter.size() * sizeof(float), m_size);
+      }
+
+      switch (checkPrecision(out, in)) {
```
Can we replace this with `instantiatePreconditioner` or similar to reduce the boiler plate and only compile which precisions are enabled?

hummingtree replied:
~~Similar to the comment in `madwf_transfer.cu`, Only float is instantiated here, and we only need float.~~

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/ff335265b55f103ee3927abdf7ef5671034dade2

maddyscientist commented on lib/madwf_tensor.cu:
```diff
@@ -0,0 +1,123 @@
+
+      size_t m_size = in.X(4) * out.X(4) * sizeof(matrix_type);
+      if (transfer_parameter.size() * sizeof(float) != m_size) {
+        errorQuda("Training Parameter size mismatch %lu neq %lu.\n", transfer_parameter.size() * sizeof(float), m_size);
```
Trailing `\n` not needed for `errorQuda`

maddyscientist commented on lib/madwf_tensor.cu:
```diff
@@ -0,0 +1,123 @@
+
+      unsigned int sharedBytesPerBlock(const TuneParam &) const { return 0; }
+
+      bool tuneGridDim() const { return false; } // Don't tune the grid dimensions.
```
This isn't necessary here.  For grid-size tuning, you would use `TunableGridStrideKernel3D`

maddyscientist commented on lib/madwf_tensor.cu:
```diff
@@ -0,0 +1,123 @@
+      {
+        constexpr int block_size = 128;
+
+        qudaEvent_t event = qudaEventCreate();
```
Event creation should not be done in the `apply` function.  Why not just remove the event usage here, and add a synchronization after the `apply`?

maddyscientist commented on lib/madwf_tensor.cu:
```diff
@@ -0,0 +1,123 @@
+        return maxDynamicSharedBytesPerBlock();
+      }
+
+      unsigned int sharedBytesPerBlock(const TuneParam &) const { return 0; }
```
Can be deleted since the parent default returns 0.

maddyscientist commented on lib/madwf_tensor.cu:
```diff
@@ -0,0 +1,123 @@
+  namespace madwf_ml
+  {
+
+    template <class storage_type, class matrix_type> class tensor_5D_wrapper : public TunableKernel3D
```
General style comment going forward.  For templated types, I think we should adopt the C++ convention of `storage_t` and `matrix_t`.

maddyscientist commented on lib/madwf_tensor.cu:
```diff
@@ -0,0 +1,123 @@
+
+      void apply(const qudaStream_t &stream)
+      {
+        constexpr int block_size = 128;
```
I'm a bit confused here.  Is block size tuning deployed in this kernel?  If the block size needs to always be 128, then this can be enabled cleaner using the `Tunable` parent methods for setting the min and max bounds for the block size.  Or through overriding the `advanceTuneParam` method.

hummingtree replied:
Here we need two kernels here: the first kernel performs reduction within each block, and the second kernel performs the inter-block reduction. The reason why these two kernels are not fused together like in `BlockReduce` is that there the number of reduction items is too large (`(Ls * 4 * Ls * 4) * 2`). The `128` is the block size of the second kernel, and the block size of the first kernel is tuned.

maddyscientist replied:
Ah, understood thanks.  We could certainly increase the number of reductions supported in a single kernel in order to support a single fused kernel.  That would be cleaner and faster.

hummingtree replied:
Having this support would solve many of the issues you have pointed out here, e.g. there is no need to have a temporary buffer, no need to have the event and wait, need to have two kernels, etc.

maddyscientist commented on lib/device_vector.cu:
```diff
@@ -0,0 +1,73 @@
+namespace quda
+{
+
+  template <class T> struct AxpbyArg : kernel_param<> {
```
The arg and the functor need to be in a separate kernel include file, else this will break on Jitify.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/8b75d45d4fd17c8c6e8e368fd783883afcdc4d67

maddyscientist commented on include/shared_memory_cache_helper.cuh:
```diff
@@ -210,6 +210,34 @@ namespace quda
 
+    /**
+       @brief Save a vector to the shared memory cache but with cutomized index and leading dimension
+       @param[in] index The customized index
```
parameter `a` is missing from the doxygen here

maddyscientist commented on include/shared_memory_cache_helper.cuh:
```diff
@@ -210,6 +210,34 @@ namespace quda
+       @param[in] index The customized index
+       @param[in] ld The leading dimension
+     */
+    __device__ inline void save_idx(int index, int ld, const T &a)
```
Wondering if we should change the ordering here to something like the below to match the overload for `save` which has optionals thread coordinates after the parameter `a`?
```c++
__device__ inline void save_idx(const T &a, int index, int ld)
```

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/d776b593adac861010d314b14523ba261ada3c14 has made this change unnecessary.

maddyscientist commented on lib/inv_cg_quda.cpp:
```diff
@@ -242,6 +244,7 @@ namespace quda {
 
     // whether to select alternative reliable updates
     bool alternative_reliable = param.use_alternative_reliable;
+    bool advanced_feature = !(param.precondition_no_advanced_feature && param.is_preconditioner);
```
As much as I love the name `advanced_feature`, it's not very descriptive what it does 😉

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/cefb2a527fa7cb1303a9e7487b9fea4f39d1bf8d

maddyscientist commented on lib/inv_pcg_quda.cpp:
```diff
@@ -10,6 +10,9 @@
 #include <invert_quda.h>
 #include <util_quda.h>
 
+#include <madwf_ml.h>
```
I'm not sure if we should be including specific reference to the MADWF preconditioner in the PCG file?  Wondering if this can refactored into a more abstract interface so this isn't necessary?

If we end up with specific preconditioner implementation details for every preconditioner we this becomes ugly fast.  Similarly, it's possible we could want to use MADWF with a different solver.  It would be good to decouple the preconditioner and solver.

hummingtree replied:
I guess we can have a preconditioner class/interface that handles the creation of the preconditioners?. And yes I also like the idea of decoupling preconditioner and solver.

maddyscientist replied:
I like the sound of a preconditioner factory 😉

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/96926239a741adb4fa5c8bab17dc6714e094af3b

maddyscientist commented on include/madwf_ml.h:
```diff
@@ -0,0 +1,372 @@
+    int train_maxiter;
+
+    // persistent buffers for reuse.
+    ColorSpinorField *forward_tmp = nullptr;
```
You could keep @mathiaswagner happy and use `unique_ptr` here 😉

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/96926239a741adb4fa5c8bab17dc6714e094af3b

maddyscientist commented on include/madwf_ml.h:
```diff
@@ -0,0 +1,372 @@
+      strcpy(param_infile, solve_param.madwf_param_infile);
+      strcpy(param_outfile, solve_param.madwf_param_outfile);
+
+      printfQuda("Launching MADWF accelerator ... \n");
```
The `printfQuda` calls in this file should be guarded behind `getVerbosity()`, so that is the user requests silent, nothing is printed.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/96926239a741adb4fa5c8bab17dc6714e094af3b

maddyscientist commented on include/madwf_ml.h:
```diff
@@ -0,0 +1,372 @@
+    }
+
+    template <class Ref, class Base>
+    double cost(const Ref &ref, Base &base, ColorSpinorField &out, const ColorSpinorField &in)
```
Is this function called in a hot loop?  If so, the allocation overhead of temporaries may be signficant.

hummingtree replied:
I thought if we use memory pool the allocation of these temporaries would be negligible?

maddyscientist replied:
The actual memory allocation is negligible, but there are other overheads associated with field creation such as zeroing out any padded region (likely not applicable here) and the host-side setup cost of setting the meta data in the class.  Regardless, as a design goal, I think it's desirable to avoid memory allocation / free in hot loops.  The memory pool allocator does lead to an overhead, and in some cases one may need to disable the pool allocator to fit on a desired number of GPUs.

maddyscientist replied:
Having thought about this a bit more, I think we can just leave the temporary creation.  In the of DWF fermions, this really will be a negligible overhead, and we probably don't need to worry about the pool off case.  If it's an issue in the future, it can always be optimized then.  Marking as resolved.

maddyscientist commented on include/madwf_ml.h:
```diff
@@ -0,0 +1,372 @@
+#pragma once
```
This is a lot of implementation code in a header.  Can most of this be moved to a .cpp file?  Moreover, if we were to use MADWF in other solvers, we'd end up with unnecessary bloat.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/96926239a741adb4fa5c8bab17dc6714e094af3b

maddyscientist commented on include/kernels/madwf_transfer.cuh:
```diff
@@ -0,0 +1,150 @@
+#include <madwf_transfer.h>
+
+namespace quda
+{
```
Some doxygen in this file would be good.  As would a brief overall description of what this kernel does.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/cefb2a527fa7cb1303a9e7487b9fea4f39d1bf8d

maddyscientist commented on include/kernels/madwf_transfer.cuh:
```diff
@@ -0,0 +1,150 @@
+        if (out.Nspin() != 4) errorQuda("nSpin = %d not support", out.Nspin());
+        if (out.Ncolor() != 3) errorQuda("nColor = %d not support", out.Ncolor());
+
+        if (!in.isNative() || !out.isNative())
```
this can be replaced with a single call to `checkNative(in, out)`

maddyscientist commented on include/kernels/madwf_transfer.cuh:
```diff
@@ -0,0 +1,150 @@
+                    static_cast<int>(out.VolumeCB() / Ls_out));
+        }
+
+        if (in.Nspin() != 4) errorQuda("nSpin = %d not support", in.Nspin());
```
typo: support*ed*

maddyscientist commented on include/kernels/madwf_transfer.cuh:
```diff
@@ -0,0 +1,150 @@
+
+        if (x_cb >= volume_4d_cb) return;
+        if (s >= Ls_out) return;
+        if (parity >= arg.nParity) return;
```
If these are the thread dimensions that are passed to the kernel, e.g., `kernel_param.threads` then this bound checking is unnecessary here and will be done already in the generic kernel

maddyscientist commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,205 @@
+        if (out.Nspin() != 4) errorQuda("nSpin = %d not support", out.Nspin());
+        if (out.Ncolor() != 3) errorQuda("nColor = %d not support", out.Ncolor());
+
+        if (!in.isNative() || !out.isNative())
```
change to `checkNative(in, out)`

maddyscientist commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,205 @@
+                    static_cast<int>(out.VolumeCB() / Ls_out));
+        }
+
+        if (in.Nspin() != 4) errorQuda("nSpin = %d not support", in.Nspin());
```
typo: support*ed*

maddyscientist commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,205 @@
+    __device__ __host__ inline void vector_tensor_matrix(SpinMatrix<real> *mp, int m_index, const WilsonVector<real> &v,
+                                                         const WilsonVector<real> &w)
+    {
+      complex<real> *p = reinterpret_cast<complex<real> *>(mp);
```
Do we need the `reinterpret_cast`?  Can we not just use the matrix accessors directly?

hummingtree replied:
We do need the `reinterpret_cast` here since we have multiple matrices and we are using a strided access pattern.

maddyscientist replied:
I was about to type this
> The concern with the `reinterpret_cast` is that it can lead to undefined behaviour, e.g., if the compiler inserts padding in the matrix.  A good test for this is often to see if the code runs ok in debug mode (`-G`).  In almost all circumstances we should `reinterpret_cast` on stack variables at all times.
But now I see when looking closer at the code that this is a global buffer and will be allocated on the heap.  So this will be safe to do, and I can see the point of this 😄

maddyscientist commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,205 @@
+          int cs = a * spin_dim + b;
+          complex<real> z = 0;
+#pragma unroll
+          for (int color = 0; color < color_dim; color++) { z += conj(conj(v(a, color)) * w(b, color)); }
```
I think you can replace this with `complex<real> z = innerProduct(v, w, a, b)`

maddyscientist commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,205 @@
+          errorQuda("Unsupported field order out=%d in=%d\n", out.FieldOrder(), in.FieldOrder());
+
+        size_t reduce_bytes = num_x_blocks * sizeof(matrix_type) * out.X(4) * in.X(4);
+        reduce_buffer = reinterpret_cast<matrix_type *>(device_malloc(reduce_bytes));
```
We should keep the argument classes trivially copyable.  So best here to move the reduce_buffer allocation and free outside of the constructor / destructor.

hummingtree replied:
The need to have the buffer has been removed.

maddyscientist commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,205 @@
+        const int volume_4d_cb = arg.volume_4d_cb;
+        auto reduce_buffer = arg.reduce_buffer;
+
+        if (x_cb >= volume_4d_cb) return;
```
Not sure if these bounds checks are needed here.

maddyscientist commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,205 @@
+        int ld = Ls_in * target::block_dim().x;
+        int t = s;
+        while (t < Ls_in) {
+          int index = t * target::block_dim().x + target::thread_idx().x;
```
Just wondering here why the standard interface for the cache load/save can't be used here?

hummingtree replied:
Since the standard load/save interface is based on the current block shape. Here the `y` dimension is based on the Ls_in instead of `block_dim().y`.

maddyscientist replied:
The block size that the interface utilizes can be set in the constructor for the `SharedMemoryCache`, so it doesn't have to match the block size.  Or does this not work in this case for some reason?

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/d776b593adac861010d314b14523ba261ada3c14

maddyscientist commented on include/kernels/madwf_transfer.cuh:
```diff
@@ -0,0 +1,150 @@
+        for (int column = 0; column < spin_dim; column++) {
+          auto v_col = v(column, color);
+#pragma unroll
+          for (int row = 0; row < spin_dim; row++) {
```
The code below won't lead to optimal FMA issuance.  Quick fix is to write this as
```c++
if (dagger)
  out(row, color) = cmac(conj(m(column, row), v_col, out(row, color));
else
  out(row, color) = cmac(m(row, column), v_col, out(row, color));
```

maddyscientist commented on include/invert_quda.h:
```diff
@@ -487,9 +540,16 @@ namespace quda {
-    const DiracMatrix &Mprecon() { return matPrecon; }
+    virtual void train_param(Solver &, ColorSpinorField &) { errorQuda("NOT implemented."); }
+
+    virtual void solve_and_collect(ColorSpinorField &, ColorSpinorField &, std::vector<ColorSpinorField *> &, int, double) { errorQuda("NOT implemented."); }
```
Documentation needed here for what `train_param` and `solve_and_collect` do.  And I'm not sure if I like having these very MADWF concepts in the solver interface.  Wondering if another approach is possible.

hummingtree replied:
This is an interface problem that I was having trouble with. Here what we want to do is collect some vectors during the solve, and this interface change that I ended up with, which is ugly and not optimal. But I don't have a better idea.

maddyscientist replied:
Perhaps this can be a discussion on our next QUDA call

weinbe2 replied:
Sticking a pin on revisiting this in our next call

weinbe2 replied:
I don't mind having a `solve_and_collect` specifically---I did something similar with `PrepareSpecialMG` or whatnot with staggered MG. I think we should write it a bit differently, with a `solveAndCollect` and a 'hasSolveAndCollect`, and then we can query for that appropriately elsewhere.

I don't mind the idea of having a `solveAndCollect` in general, who knows what ideas we may have in the future.

weinbe2 replied:
I might want something like this for DWF MG when I want to shrink `Ls`!

maddyscientist commented on include/invert_quda.h:
```diff
@@ -673,6 +733,70 @@ namespace quda {
     virtual bool hermitian() { return true; } /** CG is only for Hermitian systems */
   };
 
+  /**
```
I think this should be pulled out in its own header

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/96926239a741adb4fa5c8bab17dc6714e094af3b

maddyscientist commented on include/invert_quda.h:
```diff
@@ -673,6 +733,70 @@ namespace quda {
   };
 
+  /**
+     @brief  Generic solver with an accelerator.
```
More documentation needed here.  My naive 🧠 would assume an accelerator is a GPU 😛

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/96926239a741adb4fa5c8bab17dc6714e094af3b

maddyscientist reviewed (changes requested):
Undoubtedly great work @hummingtree 😄, but I think the PR needs a bit more refinement and cleanup for integration with the greater QUDA.  I've left a lot of comments inline for things to be addressed.

hummingtree committed:
Fix a typo in polynomial.h and add doxygen in device_vector.h

hummingtree committed:
Change from *_type to *_t to match the cpp convention.

hummingtree committed:
Address some comments in madwf_tensor/transer.cuh. Still need to address the problem of having temp memory and events in apply.

hummingtree committed:
Refactor:
- Add a separate function to handle the creation of preconditioner
- Add a accelerator.h
- Move the implementation of madwf_ml.h to madwf_ml.cpp
- Add guard to printfQuda's in madwf_ml.cpp/.h

hummingtree committed:
Use M_PI. Use the existing load/save routines for shared cache.

hummingtree committed:
Merge remote-tracking branch 'origin/develop' into feature/mspcg-ml

hummingtree committed:
Add doxygen to madwf_tensor/transfer.cuh and to the advanced feature in inv_cg_quda.cpp

hummingtree committed:
Pack the MADWF parameters into a struct.

hummingtree committed:
Move kernel and arg of device_vector.cu to a .cuh file.

weinbe2 commented on include/invert_quda.h:
```diff
@@ -791,7 +821,7 @@ namespace quda {
   class PreconCG : public Solver {
     private:
-      Solver *K;
+      std::unique_ptr<Solver> K;
```
Not necessarily a design question we need to solve here, but a `std::unique_ptr` isn't a good solution in all cases---for ex, if we pass in an external preconditioner (such as an MG preconditioner) as opposed to letting the solver create it for us.

See the two different constructors for the `GCR` solver as an example.

hummingtree replied:
So we should use `shared_ptr` here maybe? Here in PCG as its current shape `unique_ptr` and `shared_ptr` are effectively the same.

weinbe2 replied:
I think the issue with a `shared_ptr` is it depends on the `MG` preconditioner (or any external preconditioner) being held in a `shared_ptr` everywhere for reference counting reasons, which isn't the case here (though it is something we should address).

I'm inclined to say we should make it a `shared_ptr` here for forward-thinking reasons, since it's all the same in PCG for now, and we can think about making it more robust for external preconditioners in a later PR. Thoughts?

hummingtree replied:
Fair point and I am happy to change to `shared_ptr`.

hummingtree committed:
Cover the accelerator with time profile.

hummingtree committed:
Change MADWf tensor to use QUDA's reduction framework. Currently have spills so perf is bad.

hummingtree committed:
[Not Working] Change to reduce 4x4 complex<float> per thread.

hummingtree committed:
Fix the MADWF tensor code: add an optional buffer to ReduceArg so the output result can be written directly. Fix namings of the MADWF tensor code so they make more sense.

hummingtree committed:
Use shared_ptr for preconditioner. Remove used complex_t.

hummingtree committed:
Use an instantiate helper for madwf_tensor and madwf_transfer. Remove the unecessary templates. Further refactor of madwf_ml.h and madwf_transfer.h

hummingtree committed:
Add MADWF parameters to fortran.

hummingtree committed:
Merge remote-tracking branch 'origin/feature/multi-blas-misc' into feature/mspcg-ml

hummingtree committed:
Merge remote-tracking branch 'origin/feature/multi-blas-misc' into feature/mspcg-ml

hummingtree committed:
For multi-reduction do y<->z but still use block batch size of 1. Fix flops count.

maddyscientist commented on include/madwf_ml.h:
```diff
@@ -0,0 +1,138 @@
+      @brief A helper class to instantiate the precisions (half and quarter) for madwf_tensor and madwf_transfer
+    */
+    template <template <class T> class F, class... Args>
+    void instantiate_madwf(ColorSpinorField &out, const ColorSpinorField &in, Args... args)
```
Just noting that now we are C++17, so we could write this without the preprocessor using `if constexpr` (I plan to make this change to instantiate.h at some point.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/703f9d363082c78607d77ac0d576e66733264f35

weinbe2 replied:
Can we use `instantiatePreconditioner` for this?

hummingtree committed:
Merge remote-tracking branch 'origin/develop' into feature/mspcg-ml

maddyscientist commented on tests/utils/set_params.cpp:
```diff
@@ -222,16 +222,31 @@ void setInvertParam(QudaInvertParam &inv_param)
   inv_param.tol_precondition = tol_precondition;
   inv_param.maxiter_precondition = maxiter_precondition;
-  inv_param.verbosity_precondition = mg_verbosity[0];
+  inv_param.verbosity_precondition = verbosity_precondition;
```
Does this change affect the verbosity settings when running multigrid?

hummingtree replied:
This is in the `setInvertParam` so I think it is irrelevant for MG? And this is only for tests so should not affect the general behavior?

weinbe2 replied:
I wholeheartedly agree with this change, it actually caused me some confusion while working on the local version of the Schur op. If I had noticed this when it was originally introduced I would've flagged it. This is dealt with appropriately for the MG codepath iirc, where a separate `inv_param` structure is created.

maddyscientist commented on lib/madwf_tensor.cu:
```diff
@@ -0,0 +1,78 @@
+      using reduce_t = typename Arg::reduce_t;
+      reduce_t *wm_p;
+
+    private:
```
This private declaration is redundant in a class.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/703f9d363082c78607d77ac0d576e66733264f35

maddyscientist commented on lib/madwf_tensor.cu:
```diff
@@ -0,0 +1,78 @@
+        TunableMultiReduction(x, x.X(4) * y.X(4)), x(x), y(y), wm_p(reinterpret_cast<reduce_t *>(wm_p))
+      {
+        char tmp[512];
+        sprintf(tmp, ",%02d->%02d", y.X(4), x.X(4));
```
Using `sprintf` for int to string conversion is very slow.  Use `u32toa` instead (uint_to_char.h).

hummingtree replied:
But using `sprintf` for strings (i.e. `%s`) is fast enough?

maddyscientist replied:
My preference would be to not use `sprintf` at all 😉  It's probably worth looking at `std::string` instead of `char*`, and also we could look into fmtlib (https://github.com/lattice/quda/issues/1176) for conversion, since it looks to be fast and is the basis for C++20 `std::format`.  For now, as long we don't use `sprintf` for integer to string, I'm happy.

hummingtree replied:
![n9n6sqvOnVY51yQX8ciX-BTJTsw](https://user-images.githubusercontent.com/12970356/148278757-fdb2f2db-9f17-4be0-8f23-9ffeb020f8c4.png)
I did a quick benchmark of 4 different options:
- Using `sprintf("%d")` (the original one, `Sprintf_d`)
- Using `i32toa` and `sprintf("%s")` (`Sprintf_s_i32toa`)
- Using `i32toa` and `strcat` (`Strcat_i32toa`)
- Using `i32toa` and `std::string` (`StdString_i32toa`)

Seems we should avoid `sprintf`.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/703f9d363082c78607d77ac0d576e66733264f35

maddyscientist commented on lib/madwf_ml.cpp:
```diff
@@ -0,0 +1,340 @@
+
+      profile.TPSTART(QUDA_PROFILE_IO);
+
+      char param_file_name[512];
```
Maybe move this i/o handling to a separate function to improve readability?

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/703f9d363082c78607d77ac0d576e66733264f35

maddyscientist commented on lib/inv_cg_quda.cpp:
```diff
@@ -242,6 +244,14 @@ namespace quda {
+    /**
+      When CG is used as a preconditioner, we do not want enable the `advanced features`, which includes:
+      - Reliable updates
+      - Pipelining
```
Why wold we not want to use pipelining when CG is used a preconditioner?

hummingtree replied:
I should change the wording here. These features are turned off if CG is used as a preconditioner and `precondition_no_advanced_feature` is set to true. In `invert_quda.h`,
```
precondition_no_advanced_feature(param.schwarz_type == QUDA_ADDITIVE_SCHWARZ || param.schwarz_type == QUDA_ADDITIVE_MADWF_SCHWARZ)
```

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/7e090a3c13613aee3586545a9289f60982c00cbd

maddyscientist commented on include/targets/cuda/reduce_helper.h:
```diff
@@ -76,6 +76,7 @@ namespace quda
     cuda::atomic<system_atomic_t, cuda::thread_scope_system> *result_h; /** host atomic buffer */
     count_t *count; /** count array that is used to track the number of completed thread blocks at a given batch index */
     bool consumed; // check to ensure that we don't complete more than once unless we explicitly reset
+    T *device_output_buffer = nullptr; // Optional device output buffer for the reduction result
```
Changes to this file must also be propagated to include/targets/generic/reduce_helper.h (which is the stream synchronized reduction, as opposed to this which is the `std::atomic` based implementation).

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/d9e668e4f01f639c0fb01188d4fbbcd2cbf34544

maddyscientist commented on include/targets/cuda/reduce_helper.h:
```diff
@@ -209,6 +220,8 @@ namespace quda
             arg.result_d[n * idx + i].store(sum_tmp[i], cuda::std::memory_order_relaxed);
           }
+        } else if (arg.device_output_buffer) {
+          arg.get_device_output_buffer()[idx] = sum;
```
The implication here is that if the manually set device buffer is used, there is no synchronization possible since it doesn't use an atomic write.  With this in mind, I think the getter and setter functions should labelled `device_output_async_buffer` to make this more explicit, and note this in the doxygen comments for these functions.

maddyscientist replied:
And thinking about this: I'm not sure we need the name `device` in the title.  It could be a pinned host buffer for example, or if this function was run on the host (in some future code evolution), then the `device` name would be misleading.  So how about `get_output_async_buffer` and `set_output_async_buffer`?

hummingtree replied:
But then we need a way to tell if it is a device buffer or host buffer right?

maddyscientist replied:
Yes, but that logic would be handled at a higher level when one is deciding to run on host or device.  For now, since this buffer is always on the device, there is no risk of ambiguity.  And it's just a name 😄

hummingtree replied:
Sound good. I will change to `get_output_async_buffer` and `set_output_async_buffer`.

hummingtree replied:
Discussed offline and I should check if the async knob is turned on when the setter function is called.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/d9e668e4f01f639c0fb01188d4fbbcd2cbf34544

maddyscientist commented on include/polynomial.h:
```diff
@@ -0,0 +1,112 @@
+  inline std::vector<double> quadratic_formula(std::array<double, 3> coeff)
+  {
+
+    std::vector<double> z;
```
Why not just write this as `std::vector<double> z(2)`?

hummingtree replied:
See the response to the following comment.

maddyscientist commented on include/polynomial.h:
```diff
@@ -0,0 +1,112 @@
+  inline std::vector<double> cubic_formula(std::array<double, 4> coeff)
+  {
+
+    std::vector<double> t;
```
`std::vector<double> t(3);`

hummingtree replied:
The size of the vector is used to indicate the number of real roots found, which is why I use `reserve` here.

maddyscientist replied:
Sure, so why not declare the size of the vector in the constructor?

hummingtree replied:
Since the size is found later in the code so it can not be determined here.

maddyscientist replied:
Ah, understood.  Thanks

maddyscientist commented on include/madwf_ml.h:
```diff
@@ -0,0 +1,138 @@
+
+    static constexpr madwf_ml::transfer_5D_t transfer_t = madwf_ml::transfer_5D_t::Spin;
+
+    // The parameters to be trained.
```
Probably should upgrade all these `// comments` to doxygen

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/377701f828ca406411764326773a4dcfdefd8b74

maddyscientist commented on include/kernels/madwf_transfer.cuh:
```diff
@@ -0,0 +1,152 @@
+        const matrix_t *wm_p = arg.wm_p;
+
+        int thread_idx = target::thread_idx().y * target::block_dim().x + target::thread_idx().x;
+        SharedMemoryCache<real> cache;
```
Just wondering if this can be written using `SharedMemoryCache<matrix_t>` instead.  Loading in each element, real by real, will mean the compiler can't use 64-bit or 128-bit instructions.  Could the matrix be assembled first in registers, and then copied to shared memory, and perhaps this would negate needing to accessing `data()` directly?

hummingtree replied:
Each thread loading one `matrix_t` would result in only `Ls_in * Ls_out` number of loads for a threadblock, which has too little parallelism for some block size. I have tried with `SharedMemoryCache<matrix_t>` and it results in a slow down when `Ls_in = 4`. Perhaps we can keep the code in this shape this for now?

maddyscientist replied:
Ah, this makes sense.  Yep, fine to leave this as is then.

maddyscientist commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,114 @@
+        for (int y_spin = 0; y_spin < spin_dim; y_spin++) {
+#pragma unroll
+          for (int x_spin = 0; x_spin < spin_dim; x_spin++) {
+            sum[x_spin * spin_dim + y_spin] += conj(innerProduct(x, y, x_spin, y_spin));
```
Could this not simply be written as ?
```c++
sum[x_spin * spin_dim + y_spin] += innerProduct(y, x, y_spin, x_spin);
```

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/d9e668e4f01f639c0fb01188d4fbbcd2cbf34544

maddyscientist commented on include/device_vector.h:
```diff
@@ -0,0 +1,150 @@
+    {
+      size_t bytes = _size * sizeof(real);
+      if (bytes > 0) {
+        _device_data = reinterpret_cast<real *>(device_malloc(bytes));
```
This should probably be `pool_device_malloc` ?

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/377701f828ca406411764326773a4dcfdefd8b74

maddyscientist commented on include/device_vector.h:
```diff
@@ -0,0 +1,150 @@
+    @brief This class `device_vector` provides a light weighted wrapper for data on device.
+  */
+
+  template <class real> class device_vector
```
I'm not going to object to this addition, as it seems very useful, but the addition of an object named `device_vector` seems antithetical to the current drive to unify objects, e.g., `cudaColorspinorField` / `cpuColorspinorField`.  I reserve the right to revisit this in the future 😜

maddyscientist commented on include/accelerator.h:
```diff
@@ -0,0 +1,80 @@
+    - Solvers are used in iterations based on an operator and a vector, e.g. to solver A * x = b, we have
+        b = Solver[A, x].
+    - Now assume we have another set of operator B and vector c and y. The acceleration works as
+        y = forward(x) -> c = Solver[B, y] -> b = backward(c),
```
Perhaps we should make the analogy with multigrid language here explicit.  E.g., `forward` is the restrictor, `backward` is the prolongator and `Solver[B, ...]` is the coarse grid solve.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/377701f828ca406411764326773a4dcfdefd8b74

hummingtree committed:
Address some of PR comments:
- Use if constexpr for the instantiations for the precisions
- Remove redundant private in classes
- Get rid of sprintf's since they are slow
- Move I/O handling in MADWF training to separate functions

hummingtree committed:
- Add analogy with MG.
- Move some of the comments to the class members to doxygen
- Use pool_device_malloc instead of device_malloc

hummingtree committed:
- Change from conj(innerProduct(x, y, x_spin, y_spin)) to innerProduct(y, x, y_spin, x_spin)
- Add "async" to the reduction output buffer names to make it clear they are used for async ops
- Repeat what's been done in cuda/reduce_helper.cuh to generic/reduce_helper.cuh

hummingtree committed:
Change the wordings for CG for the advanced features.

hummingtree committed:
Use pool_device_free. Use static_cast<bool>(PRECISION & 2).

hummingtree committed:
Change all device_malloc to pool_device_malloc, and device_free to pool_device_free.

hummingtree committed:
Remove unnecessary error check for malloc.

hummingtree commented:
> This error check is superfluous, since the allocation will be checked inside the malloc function.

I have removed those in https://github.com/lattice/quda/pull/1061/commits/255ec92493150a420f9fad3ce70d09f6675eaa94.

hummingtree committed:
Merge remote-tracking branch 'origin/develop' into feature/mspcg-ml

weinbe2 commented on include/kernels/madwf_transfer.cuh:
```diff
@@ -0,0 +1,152 @@
+
+  namespace madwf_ml
+  {
+    constexpr int spin_dim = 4;
```
I think we should use the variable conventions we use elsewhere of `nSpin`, `nColor`, etc, as opposed to `spin_dim`, etc. I don't necessarily have a good idea for `color_spin_dim` or `wm_dim`, though. What does `wm` stand for?

weinbe2 replied:
I also think it's a bit weird they're floating around outside of a class, couldn't `nColor`, etc, be template parameters to the various classes like they are elsewhere in the code?

hummingtree replied:
`wm` is the dimension of a Wilson matrix. `sm` means the dimension of a spin matrix. Let me try to use `nSpin` and `nColor`, and also see if I can make things simpler without using `wm` or `sm`.

hummingtree replied:
About `nColor` being a template parameter: here we are only intending to have `nColor = 3`. The other option is to have a `nColor` template and add a `static_assert` to make sure it is `3`, just it seems redundant to me.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/322c1fea1cac8aa32988f4cee0a5e5b51c628344. I have made the corresponding changes.

hummingtree committed:
Use nSpin instead of spin_dim. Use nSpin and nColor as tempalte parameters.

weinbe2 commented on include/device_vector.h:
```diff
@@ -0,0 +1,148 @@
+    @brief This class `device_vector` provides a light weighted wrapper for data on device.
+  */
+
+  template <class real> class device_vector
```
A request for a global replace---I think this should be `DeviceVector`, not `device_vector`, to match otherwise ubiquitous naming conventions (`ColorSpinorField`, `GaugeField`, `SharedMemoryCache`, ...)

hummingtree replied:
Here it was originally meant to mimic `thrust::device_vector`. Are we sure we want to make this change?

maddyscientist replied:
I think we should stick with `device_vector`

weinbe2 replied:
Independent of the motivation I still think it's weird for it to be the one "memory container" that goes against our other style conventions, but I'll let it be.

hummingtree committed:
Merge remote-tracking branch 'origin/develop' into feature/mspcg-ml

weinbe2 commented on include/reliable_updates.h:
```diff
@@ -0,0 +1,215 @@
+#include <blas_quda.h>
+#include <util_quda.h>
+
+struct reliable_updates {
```
Style convention: I think this should be `ReliableUpdates`

weinbe2 replied:
Otherwise---I really like how reliable updates have been abstracted into a separate `struct`.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/aca1ee4c94659aab2be39a304cdd6c32c6408ab9

weinbe2 commented on include/reliable_updates.h:
```diff
@@ -0,0 +1,215 @@
+
+  int rUpdate = 0;
+
+  reliable_updates(bool alternative_reliable, double u, double uhigh, double Anorm, double r2, double delta,
```
This constructor is a bit painful. I think, to match other design patterns, we should have a `ReliableUpdatesParams` struct that then gets passed to `reliable_updates` (or `ReliableUpdates` as I suggested :) )

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/aca1ee4c94659aab2be39a304cdd6c32c6408ab9

weinbe2 commented on include/invert_preconditioner.h:
```diff
@@ -0,0 +1,37 @@
+
+namespace quda
+{
+  std::shared_ptr<Solver> create_preconditioner(const DiracMatrix &matPrecon, const DiracMatrix &matEig,
```
Maybe outside the scope of this PR, but I think we should create a unified `interface` for when we have to create a preconditioner (what we do here) as opposed to when an existing one gets handed in (for multigrid --- see the two constructors for `GCR`). This would let us clean up some hackish boilerplate in the solver factory.

If anything we should make an issue for this---let me know if you think we should punt this to a separate PR (my MG-PCG work).

hummingtree replied:
See issue #1240.

weinbe2 commented on include/enum_quda.h:
```diff
@@ -182,8 +182,9 @@ typedef enum QudaMultigridCycleType_s {
-  QUDA_ADDITIVE_SCHWARZ,
-  QUDA_MULTIPLICATIVE_SCHWARZ,
+  QUDA_ADDITIVE_SCHWARZ = 0,
+  QUDA_ADDITIVE_MADWF_SCHWARZ = 1,
```
I don't like taking advantage of `QudaSchwarzType` to distinguish between additive schwarz and MADWF schwarz; we could use a separate enum in a separate field that's MDWF-specific.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/5d9e47921fe9b8a5bcb89857ef843062baaba806

weinbe2 commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,117 @@
+        volume_4d_cb(y.VolumeCB() / y.X(4)),
+        nParity(y.SiteSubset())
+      {
+        if (volume_4d_cb != static_cast<int>(x.VolumeCB() / x.X(4))) {
```
Do we need this check for both `x` and `y`?

hummingtree replied:
In the initialization list `volume_4d_cb(y.VolumeCB() / y.X(4))` so we only need to check for `x`.

weinbe2 commented on include/kernels/madwf_tensor.cuh:
```diff
@@ -0,0 +1,117 @@
+      const F x; // xput vector field
+      const F y; // yput vector field
+
+      const int_fastdiv Ls_x; // length of 5th dimension of the x field
```
Doesn't need to be an `int_fastdiv`

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/65dc02461126ad5aee630ce9153055476d15d98b. We don't actually need this variable.

weinbe2 commented on include/madwf_ml.h:
```diff
@@ -0,0 +1,154 @@
+    @param host_training_param_cache a static map that holds the cached parameters, so we do not need to load the same
+      set of parameters multiple times from disk
+  */
+  struct MadwfAcc {
```
Not for this PR, but I do like this abstraction of a "trainable accelerator". In the future we could create an (abstract?) `Acc` class that other accelerators (like this one) inherit from. We could just file the idea as an issue, I don't know if it's worth introducing a generic "trainable" abstraction before we have more than one.

weinbe2 replied:
Update on this thought: I didn't realize `MadwfAcc` was explicitly added to `inv_pcg_quda.cpp`. That makes me want to create the above abstraction sooner as opposed to later so we have `Acc::[...]` in the solver as opposed to `MadwfAcc`.

hummingtree replied:
I should be able to move the explicit `MadwfAcc` stuff out of `PCG` so we could delay this until we have another trainable accelerator?

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/3aa0eb3e4790a16d47e77065f73969f4100835b5

weinbe2 commented on include/madwf_ml.h:
```diff
@@ -0,0 +1,154 @@
+      @param[in] Ls_base the reduced 5th dimension size
+      @param[in] param_size the expected parameter size
+    */
+    void load_parameter(int Ls, int Ls_base, size_t param_size);
```
Why do we need to pass in the `param_size`? Shouldn't the loading routine know exactly how large it should be in advance?

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/7ef32163893b0cab221e561ff27c28c696576277

weinbe2 commented on lib/inv_pcg_quda.cpp:
```diff
@@ -75,41 +78,36 @@ namespace quda
+  void PreconCG::solve_and_collect(ColorSpinorField &x, ColorSpinorField &b, std::vector<ColorSpinorField *> &v_r,
+                            int collect_miniter, double collect_tol)
   {
+    K->train_param(*this, b);
```
Looping back to my idea of having a `hasSolveAndCollect` --- we could create a `if ([...]->hasSolveAndCollect) K->train_param` pattern in `operator()` itself, avoiding a separate `operator()` and `solve_and_collect`. We'd have to think about how to pass in `v_r`, `collect_miniter`, and `collect_tol` --- maybe those can become members of `Solver` themselves, at least the last two?

hummingtree replied:
I believe per our discussion offline, we will keep the virtual member function `solve_and_collect`, but make `SchwarzType` and `MADWF` separate?

hummingtree replied:
See https://github.com/lattice/quda/pull/1061/commits/5d9e47921fe9b8a5bcb89857ef843062baaba806.

weinbe2 commented on lib/inv_pcg_quda.cpp:
```diff
@@ -331,49 +349,42 @@ namespace quda
           rMinvr = reDotProduct(rSloppy, *minvrSloppy);
-          beta = rMinvr / rMinvr_old;
 
+          beta = (rMinvr - r_new_Minvr_old) / rMinvr_old;
```
I'm guessing this is because we didn't have Polak-Ribière in PCG before? (Which means I misread PCG in the past!)

hummingtree replied:
I believe that is the case.

hummingtree replied:
@weinbe2 Looking at this again, we did have Polak-Ribière in PCG previously, but only when reliable update is not triggered. The code path here is the one where reliable update is performed.

weinbe2 commented on lib/inv_pcg_quda.cpp:
```diff
@@ -443,4 +457,6 @@ namespace quda
     return;
   }
 
+  std::unordered_map<std::string, std::vector<float>> MadwfAcc::host_training_param_cache; // empty map
```
Having something like this in `PCG` implies that maybe we should introduce the `Acc` class abstraction, then a `MadwfAcc` that inherits it, sooner as opposed to later.

hummingtree replied:
I should move this to `madwf_ml.cpp` - I should have moved this there when I create `madwf_ml.cpp`.

weinbe2 commented on lib/madwf_ml.cpp:
```diff
@@ -0,0 +1,347 @@
+    ColorSpinorField null_x(csParam);
+    ColorSpinorField null_b(csParam);
+
+    RNG rng(null_b, 1234);
```
Maybe orthogonal to this PR, but we use `1234` as the seed in _so many_ places in QUDA. I think we should at least use a different number here, and in the future make the seed something we specify externally and pass in.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/65dc02461126ad5aee630ce9153055476d15d98b

weinbe2 commented on include/madwf_ml.h:
```diff
@@ -0,0 +1,154 @@
+  */
+  struct MadwfAcc {
+
+    using transfer_float = float;
```
~Is this abstraction necessary? More specifically, can it just be `float`, or is there a perf penalty of note to just having it be `double`?~

Disregard, I misunderstood how this was used.

weinbe2 commented on lib/madwf_ml.cpp:
```diff
@@ -0,0 +1,347 @@
+
+    std::string save_param_path(param.madwf_param_outfile);
+    char cstring[512];
+    sprintf(cstring, "/madwf_trained_param_rank_%05d_ls_%02d_%02d_mu_%.3f.dat", comm_rank(), Ls, Ls_base, mu);
```
This should also encode the decomposition (gridsize) to be safe. I could imagine someone copying these files from, say, Summit to Perlmutter, where you'd use a different decomposition.

hummingtree replied:
In principle the parameters here should be independent of the decomposition (it is purely a 5th dimension thing), so for the same lattice one should be able to use the same parameter for different decomposition. But in practice the decomposition could matter. I guess the correct thing to do is allowing user to add a customized string so they have the control over this?

weinbe2 replied:
Ah, okay, so I misunderstood---I thought there were different parameters for different domains. In that case, I'm guessing only one rank should be responsible for load/save, and the `load` in particular should broadcast to other ranks.

weinbe2 replied:
Otherwise I'm not sure why each rank needs to write out its own file

hummingtree replied:
Per our discussion, for now, we/I should change the code so that it only runs training on one rank, and at the end broadcasts the trained parameter to all ranks.

hummingtree replied:
Well thought about it for a while, only running training on rank zero is fine, but if it is not rank zero, it would conflict with the fact that only rank zero runs tuning for non-dslash kernels. I think for now we should still run training on all ranks, but only write out one parameter to the disk?

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/60e2f6d2f499192c4eb9dd4812a22c8e76bfd6cd

weinbe2 commented on lib/madwf_ml.cpp:
```diff
@@ -0,0 +1,347 @@
+
+    char param_file_name[512];
+    // Note that all ranks load from the same file.
+    sprintf(param_file_name, "/madwf_trained_param_rank_%05d_ls_%02d_%02d_mu_%.3f.dat", 0, Ls, Ls_base, mu);
```
See comment on encoding decomposition above

weinbe2 commented on lib/madwf_ml.cpp:
```diff
@@ -0,0 +1,347 @@
+    char cstring[512];
+    sprintf(cstring, "/madwf_trained_param_rank_%05d_ls_%02d_%02d_mu_%.3f.dat", comm_rank(), Ls, Ls_base, mu);
+    save_param_path += std::string(cstring);
+    FILE *fp = fopen(save_param_path.c_str(), "w");
```
We should check if there's an error on the open before trying to write

weinbe2 commented:
This PR overall looks good, and the majority of my comments are related to naming/design/abstractions.

One blanket comment I have is there's a general lack of Doxygen, but we should get the last mile of design discussions out of the way first.

hummingtree committed:
Move host_training_param_cache to madwf_ml.cpp

hummingtree committed:
Address a few of the PR comments:
- Remove Ls_x: it is not used.
- Move the profiles a bit so all parts are covered.
- In save_parameter check to see if the file is actually opened.

hummingtree committed:
Let load_parameter figure out the parameter size by itself.

hummingtree committed:
Only rank zero write parameter out. When training, broadcast rank zero's parameter.

hummingtree commented on include/kernels/madwf_transfer.cuh:
```diff
@@ -0,0 +1,145 @@
+#pragma unroll
+          for (int row = 0; row < nSpin; row++) {
+            if (dagger) {
+              out(row, color) = cmac(conj(m(column, row)), v_col, out(row, color));
```
Need to use `cmac` to generate optimal code.

hummingtree replied:
I must be drunk or something - these are already written `cmac` it seems. @maddyscientist

maddyscientist replied:
I think I must be drunk too. I could have sworn these weren’t `cmac` when looking earlier 🤣

hummingtree commented on tests/utils/command_line_params.cpp:
```diff
@@ -498,6 +518,21 @@ std::shared_ptr<QUDAApp> make_app(std::string app_description, std::string app_n
     ->add_option("--precon-schwarz-type", precon_schwarz_type,
                  "The type of Schwarz preconditioning to use (default=invalid)")
     ->transform(CLI::QUDACheckedTransformer(schwarz_type_map));
+
```
Move these `MADWF` parameters to a separatae CLI group.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/5d9e47921fe9b8a5bcb89857ef843062baaba806

hummingtree committed:
Create a QudaAcceleratorType enum to dictate what accelerator to use. Move MADWF parameters to a separate group.

hummingtree committed:
Move the parameters of reliable_updates to a params struct. Change to ReliableUpdates.

hummingtree committed:
In madwf tensor and transfer, check if the in/out fields are 5D fields.

hummingtree committed:
Add more checks for 5D fields.

hummingtree commented:
@weinbe2 I believe the code should now be in a state where you can merge it into your PCG PR. If one run with the MADWF accelerator (`--precon-accelerator-type madwf`), the underlying color spinor fields should be 5th dimensional fields. (Even if not with Mobius fermion, as long as the fields are 5 dimensional, the code should run: whether or not the train would succeed is another question.) I have added a check so an error would be given if the input fields are not 5 dimensional.

maddyscientist committed:
optimization of transfer_5d

maddyscientist reviewed (approved):
I've done some tests, and I'm very happy with this.  Approving (subject to addressing sign off from @weinbe2 and application of `clang-format`).  Great stuff @hummingtree.

weinbe2 commented on lib/madwf_ml.cpp:
```diff
@@ -79,17 +84,16 @@ namespace quda
 
       forward_tmp = std::make_unique<ColorSpinorField>(csParam);
       backward_tmp = std::make_unique<ColorSpinorField>(csParam);
+      profile.TPSTOP(QUDA_PROFILE_IO);
```
Do we need to restart `QUDA_PROFILE_INIT` here? As in, add `profile.TPSTART(QUDA_PROFILE_INIT);`?

hummingtree replied:
We `return` after this? `QUDA_PROFILE_INIT` will be called in the main solver.

weinbe2 replied:
Ah, I better see what's going on, I overlooked how `QUDA_PROFILE_INIT` was started at the top of this function. Disregard!

weinbe2 commented on lib/madwf_transfer.cu:
```diff
@@ -69,6 +69,8 @@ namespace quda
         errorQuda("ColorSpinorFields are not single parity: in = %d, out = %d", in.SiteSubset(), out.SiteSubset());
       }
 
+      if (out.Ndim() != 5 || in.Ndim() != 5) { errorQuda("we need a 5 dimensional field for this."); }
```
I _think_ we also need to check and make sure `Ncolor == 3` and `Nspin == 4`

hummingtree replied:
I have added these checks in the `madwf_tensor/transfer.cu`, instead of in the corresponding `.cuh` files in https://github.com/lattice/quda/pull/1061/commits/88818f3007b5212873c44c3482a6bf2a5e67c9c3,

weinbe2 commented on lib/madwf_ml.cpp:
```diff
@@ -0,0 +1,358 @@
+
+    // Broadcast the trained parameters
+    host_param = device_param.to_host();
+    comm_broadcast(host_param.data(), host_param.size() * sizeof(transfer_float));
```
Should this (and surrounding code) be timed?

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/88818f3007b5212873c44c3482a6bf2a5e67c9c3

hummingtree committed:
- Use std::vector<ColorSpinorField> instead of std::vector<ColorSpinorField *>.
- Move the spin and color checks to .cu files.
- Move around the profiles so everything is coverd.

hummingtree committed:
Merge remote-tracking branch 'origin/develop' into feature/mspcg-ml

hummingtree committed:
Add doxygen to invert_preconditioner.h, reliable_updates.h, and madwf_transfer.h

hummingtree committed:
Revert back to use std::vector<ColorSpinorField *>.

hummingtree committed:
Abstract the delayed x update (x updates in batch) code into a struct. Add this strategy to PCG.

hummingtree committed:
Remove dead comments. Add pragma once.

hummingtree committed:
Set the comms_flag for diracPre when schwarz type is invalid. Add commGlobal according to param in CA-CG.

hummingtree committed:
Disable the Ls check for now. Let the precondition solver print out.

weinbe2 committed:
Bugfix and clean-up in CA-CG

hummingtree committed:
Add doxygen for the methods added to solver classes. Remove the QUDA_SILENT push's. Fix the checkDWF's for the different DWF types.

weinbe2 commented on include/kernels/madwf_transfer.cuh:
```diff
@@ -0,0 +1,137 @@
+      static constexpr int nSpin = nSpin_;
+      static constexpr int nColor = nColor_;
+
+      using F = typename colorspinor_mapper<storage_t, 4, 3>::type;
```
`4, 3` -> `nSpin, nColor`, or should we have an assert that `nSpin == 4`, `nColor == 3`?

hummingtree replied:
Good catch. The assert has already been performed in the corresponding `.cu` files.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/5cce447990c8ea085125244fea58154e4271dc8a

weinbe2 commented on lib/inv_cg_quda.cpp:
```diff
@@ -361,7 +348,7 @@ namespace quda {
     // compute initial residual
     double r2 = 0.0;
-    if (param.use_init_guess == QUDA_USE_INIT_GUESS_YES) {
+    if (advanced_feature && param.use_init_guess == QUDA_USE_INIT_GUESS_YES) {
```
Why does this happen only if we're using an `advanced_feature`?

hummingtree replied:
Since I want to make sure the schwarz preconditioner does not use initial guess ...

weinbe2 replied:
Then shouldn't it be `!advanced_feature`?

hummingtree replied:
Normally `advanced_feature` would be `true`, and for schwarz it will be `false`.
```
    bool advanced_feature = !(param.precondition_no_advanced_feature && param.is_preconditioner);
```

weinbe2 replied:
Understood, thanks

weinbe2 commented on lib/madwf_ml.cpp:
```diff
@@ -0,0 +1,348 @@
+    save_param_path += std::string(cstring);
+    FILE *fp = fopen(save_param_path.c_str(), "w");
+    if (!fp) { errorQuda("Unable to open file %s\n", save_param_path.c_str()); }
+    size_t fwrite_count = fwrite(host_param.data(), sizeof(transfer_float), host_param.size(), fp);
```
I'm a little confused---is every rank now writing to the same exact file? Now that the filename is independent of the rank? That seems like trouble...

hummingtree replied:
Only rank 0 will call the `save_parameter` function.

weinbe2 replied:
I think it'd be safer to put the check within `save_parameter` as opposed to externally, or at least put a check inside the function that it's being called by rank 0.

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/5cce447990c8ea085125244fea58154e4271dc8a

weinbe2 commented on lib/madwf_ml.cpp:
```diff
@@ -0,0 +1,348 @@
+
+    char param_file_name[512];
+    // Note that all ranks load from the same file.
+    sprintf(param_file_name, "/madwf_trained_param_ls_%02d_%02d_mu_%.3f.dat", Ls, Ls_base, mu);
```
See comment above about each rank reading from the same file (admittedly, this is less dangerous than the write)

hummingtree replied:
And yes here each rank will read from the same file.

weinbe2 commented on lib/madwf_transfer.cu:
```diff
@@ -0,0 +1,94 @@
+        strcat(aux, ",");
+        i32toa(in_str, in.X(4));
+        strcat(aux, in_str);
+        strcat(aux, "->");
```
I think I'd prefer a `,` instead of an `->` here, just because that's a convention we use elsewhere, even in MG

hummingtree replied:
https://github.com/lattice/quda/pull/1061/commits/5cce447990c8ea085125244fea58154e4271dc8a

weinbe2 commented on tests/utils/host_utils.cpp:
```diff
@@ -238,7 +238,7 @@ void constructWilsonTestSpinorParam(quda::ColorSpinorParam *cs_param, const Quda
 
 void constructRandomSpinorSource(void *v, int nSpin, int nColor, QudaPrecision precision, QudaSolutionType sol_type,
-                                 const int *const x, quda::RNG &rng)
+                                 const int *const x, int nDim, quda::RNG &rng)
```
So this is something that we should file an issue for---for DWF, it's not that meaningful for a source to have non-zero elements anywhere but on the chiral walls. I guess for tests that's less consequential, but we should be careful about that...

hummingtree replied:
With this change (without assuming a 4D field) the random field is non-zero for all `s`.

weinbe2 replied:
Yes, I understand that---my point is that generally when we need a _physical_ random source for DWF it needs to live on the chiral walls. It's fine as is for now, though, for correctness checks and whatnot.

hummingtree replied:
Ah gotcha. We can have a separate function, or just a separate knob to handle that in the future.

hummingtree committed:
Address PR comments:
- "->" to ","
- Use nSpin and nColor instead of 4 and 3.
- Add the check so that only rank zero writes to disk.

hummingtree committed:
Apply clang-format.

weinbe2 reviewed (approved):
I just had a chance to get hands on with the PR, and I can confirm everything looks great. Awesome job @hummingtree ; once Jenkins completes this will be good to merge.

hummingtree committed:
Remove the debugging printf's that got left over ...

weinbe2 commented on lib/inv_ca_cg.cpp:
```diff
@@ -36,26 +36,38 @@ namespace quda {
                          // param.precision == param.precision_sloppy &&
                          // param.use_init_guess == QUDA_USE_INIT_GUESS_NO);
 
+      // printf("5\n");
```
Looks like something went awry with branches and `clang-format`...

hummingtree replied:
It was not `clang-format`'s fault. I did not clean up the debug `printf`'s before `clang-format` and did a `git commit -a` afterwards ...

hummingtree committed:
... and add in the comment that I accidently removed ...
